src/FFmpegVideoEncoder.cs

```diff
@@ -133,20 +134,48 @@ private AVPixelFormat GetAVPixelFormat(VideoPixelFormatsEnum pixelFormat)
             }
         }
 
+        public VideoFormat Format { get; set; }
+        private static double AdjustBppForComplexity(double baseBpp, double complexityFactor)
+        {
+            // Clamp complexityFactor between 0.8 and 1.2 to prevent extreme adjustments
+            complexityFactor = Math.Max(0.8, Math.Min(complexityFactor, 1.2));
+            return baseBpp * complexityFactor;
+        }
+        private void CalculateVp8BitrateParameters(
+            int width,
+            int height,
+            double frameRate,
+            double baseBpp,
+            double complexityFactor = 1.0)
+        {
+            double adjustedBpp = AdjustBppForComplexity(baseBpp, complexityFactor);
+            long pixelsPerFrame = (long)width * height;
+            long bitRate = (long)(pixelsPerFrame * adjustedBpp * frameRate);
+            _encoderContext->bit_rate = bitRate;
+
+            int bitRateTolerance = (int)(bitRate * 0.10);
+            _encoderContext->bit_rate_tolerance = bitRateTolerance;
+
+            long rcMinRate = (long)(bitRate * 0.80);
+            long rcMaxRate = (long)(bitRate * 1.20);
+            _encoderContext->rc_min_rate = rcMinRate;
+            _encoderContext->rc_max_rate = rcMaxRate;
+        }

```

Comment:
You can integrate the results with the existing `SetBitrate` method.
https://github.com/sipsorcery-org/SIPSorceryMedia.FFmpeg/blob/ef21fcf5f0f68b590294d9b8cc82f4799debc450/src/FFmpegVideoEncoder.cs#L259

Also will need to put notes on how it seems to be mutually exclusive with
https://github.com/sipsorcery-org/SIPSorceryMedia.FFmpeg/blob/ef21fcf5f0f68b590294d9b8cc82f4799debc450/src/FFmpegVideoSource.cs#L101

Reply:
Using VP8 here with the original code I just got a garbled mess of blocks - my version of this library is heavily modified and i added this code to fix issues i was having. I'm not using this library at all any more I just thought i'd contribute some of my updates back to it. Feel free to ignore.

Reply:
ah okay, this looks like it can be a quick PR, do you mind if I refactor some parts and make another PR based on this PR?

Reply:
absolutely - do what you want with it

Reply:
@ha-ves I added some changes locally to further improve matters with VP8

```
else if ((_codecID == AVCodecID.AV_CODEC_ID_VP8) || (_codecID == AVCodecID.AV_CODEC_ID_VP9))
 {
    av_opt_set(_encoderContext->priv_data, "quality", "realtime", 0);
    av_opt_set_int(_encoderContext->priv_data, "cpu-used", 8, 0); //most important
    av_opt_set(_encoderContext->priv_data, "deadline", "realtime", 0);
    _encoderContext->thread_count = calculate_optimal_thread_count(width, height);
    CalculateVp8BitrateParameters(width, height, fps, 0.1);
}

 private static int calculate_optimal_thread_count(int width, int height)
 {
     int cpu_count = av_cpu_count();
     int pixels = width * height;

     if (pixels <= 640 * 480)
     {
         return 1;
     }
     else if (pixels <= 1280 * 720)
     {
         return Math.Min(cpu_count, 2);
     }
     else if (pixels <= 1920 * 1080)
     {
         return Math.Min(cpu_count, 4);
     }
     else
     {
         return Math.Min(cpu_count, 8);
     }
 }
 
```